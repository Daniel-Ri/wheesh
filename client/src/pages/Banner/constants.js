export const SET_BANNERS = 'Banner/SET_BANNERS';
export const GET_BANNERS = 'Banner/GET_BANNERS';
export const DELETE_BANNER = 'Banner/DELETE_BANNER';
export const REMOVE_BANNER = 'Banner/REMOVE_BANNER';
