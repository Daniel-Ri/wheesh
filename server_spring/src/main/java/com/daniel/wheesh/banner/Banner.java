package com.daniel.wheesh.banner;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.util.Date;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@EntityListeners(AuditingEntityListener.class)
@Table(name = "\"Banners\"")
public class Banner {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    public Long id;

    @Column(nullable = false, name = "\"imageDesktop\"")
    private String imageDesktop;

    @Column(nullable = false, name = "\"imageMobile\"")
    private String imageMobile;

    @CreatedDate
    @Column(nullable = false, name = "\"createdAt\"")
    public Date createdAt;

    @LastModifiedDate
    @Column(nullable = false, name = "\"updatedAt\"")
    public Date updatedAt;
}
