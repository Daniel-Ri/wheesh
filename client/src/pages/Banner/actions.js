import { DELETE_BANNER, GET_BANNERS, REMOVE_BANNER, SET_BANNERS } from './constants';

export const setBanners = (banners) => ({
  type: SET_BANNERS,
  banners,
});

export const getBanners = () => ({
  type: GET_BANNERS,
});

export const deleteBanner = (bannerId, handleSuccess, handleError) => ({
  type: DELETE_BANNER,
  bannerId,
  handleSuccess,
  handleError,
});

export const removeBanner = (bannerId) => ({
  type: REMOVE_BANNER,
  bannerId,
});
