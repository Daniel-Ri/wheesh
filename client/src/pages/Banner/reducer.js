import { produce } from 'immer';
import { REMOVE_BANNER, SET_BANNERS } from './constants';

export const initialState = {
  banners: [],
};

export const storedKey = [];

const bannerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_BANNERS:
        draft.banners = action.banners;
        break;
      case REMOVE_BANNER:
        draft.banners = draft.banners.filter((banner) => banner.id !== action.bannerId);
        break;
    }
  });

export default bannerReducer;
