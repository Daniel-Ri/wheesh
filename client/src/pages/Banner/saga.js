import { takeLatest, call, put } from 'redux-saga/effects';
import { deleteBanner, getAllBanners } from '@domain/api';
import { DELETE_BANNER, GET_BANNERS } from './constants';
import { removeBanner, setBanners } from './actions';

function* doGetBanners() {
  try {
    const response = yield call(getAllBanners);
    yield put(setBanners(response.data));
  } catch (error) {
    // error
  }
}

function* doDeleteBanner({ bannerId, handleSuccess, handleError }) {
  try {
    yield call(deleteBanner, bannerId);
    yield put(removeBanner(bannerId));
    yield call(handleSuccess);
  } catch (error) {
    // error
    yield call(handleError, error.response.data.message);
  }
}

export default function* bannerSaga() {
  yield takeLatest(GET_BANNERS, doGetBanners);
  yield takeLatest(DELETE_BANNER, doDeleteBanner);
}
