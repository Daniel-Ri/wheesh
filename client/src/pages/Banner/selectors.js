import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectBannerState = (state) => state.banner || initialState;

export const selectBanners = createSelector(selectBannerState, (state) => state.banners);
