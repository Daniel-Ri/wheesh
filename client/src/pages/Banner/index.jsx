import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';

import { useMediaQuery } from 'react-responsive';
import AddOutlinedIcon from '@mui/icons-material/AddOutlined';

import BackBtn from '@components/BackBtn';
import { useEffect } from 'react';

import { useNavigate } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';
import { selectUser } from '@containers/Client/selectors';
import toast from 'react-hot-toast';
import { Button } from '@mui/material';
import { getBanners } from './actions';
import { selectBanners } from './selectors';
import BannerRow from './components/BannerRow';

import classes from './style.module.scss';

const Banner = ({ user, banners, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const isMobile = useMediaQuery({ query: '(max-width: 768px)' });

  useEffect(() => {
    dispatch(getBanners());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (user?.role !== 'admin') {
      toast.error(formatMessage({ id: 'app_not_authorized' }));
      navigate('/');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user]);

  return (
    <main className={classes.main}>
      <div className={classes.container}>
        <header>
          <BackBtn handleClickBack={() => navigate('/me')} />
          <h1>
            <FormattedMessage id="app_banner_management" />
          </h1>
        </header>

        <div className={classes.button}>
          <Button variant="contained" className={classes.btn} onClick={() => navigate('/addBanner')}>
            <AddOutlinedIcon className={classes.icon} />
            <div className={classes.message}>
              <FormattedMessage id="app_add" />
            </div>
          </Button>
        </div>

        <table className={classes.banners}>
          {!isMobile ? (
            <tbody>
              <tr>
                <th>Desktop</th>
                <th>Mobile</th>
                {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
                <th />
              </tr>
              {banners.map((banner) => (
                <BannerRow key={banner.id} banner={banner} />
              ))}
            </tbody>
          ) : (
            <tbody>
              {banners.map((banner) => (
                <BannerRow key={banner.id} banner={banner} />
              ))}
            </tbody>
          )}
        </table>
      </div>
    </main>
  );
};

Banner.propTypes = {
  user: PropTypes.object,
  banners: PropTypes.array,
  intl: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  user: selectUser,
  banners: selectBanners,
});

export default injectIntl(connect(mapStateToProps)(Banner));
