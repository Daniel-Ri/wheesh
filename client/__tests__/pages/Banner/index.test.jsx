import { render } from '@utils/testHelper';
import Banner from '@pages/Banner';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  connect: jest.fn((mapStateToProps) => (Component) => (props) => <Component {...props} {...mapStateToProps} />),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

let wrapper;

const mockProps = {
  user: {
    id: 1,
    role: 'admin',
  },
  banners: [
    {
      id: 1,
      imageDesktop: 'public/Frame 1 - Desktop.png',
      imageMobile: 'public/Frame 1 - Mobile.png',
    },
    {
      id: 2,
      imageDesktop: 'public/Frame 2 - Desktop.png',
      imageMobile: 'public/Frame 2 - Mobile.png',
    },
    {
      id: 3,
      imageDesktop: 'public/Frame 3 - Desktop.png',
      imageMobile: 'public/Frame 3 - Mobile.png',
    },
    {
      id: 4,
      imageDesktop: 'public/Frame 4 - Desktop.png',
      imageMobile: 'public/Frame 4 - Mobile.png',
    },
    {
      id: 5,
      imageDesktop: 'public/Frame 5 - Desktop.png',
      imageMobile: 'public/Frame 5 - Mobile.png',
    },
  ],
};

beforeEach(() => {
  wrapper = render(<Banner user={mockProps.user} banners={mockProps.banners} />);
});

describe('Pages Banner', () => {
  test('Renders Banner page', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('Banner')).toBeInTheDocument();
  });
});

// // pages/Banner/index.test.jsx
// import { Provider } from 'react-redux';
// import { IntlProvider } from 'react-intl';
// import { render, screen, fireEvent } from '@utils/testHelper'; // Assuming your test helper is in the specified location
// import Banner from '@pages/Banner';
// import store from '../../../src/configureStore'; // Provide the correct path to your configureStore

// // Mock the react-redux module
// jest.mock('react-redux', () => ({
//   ...jest.requireActual('react-redux'), // Use the actual implementation for most functions
//   connect: jest.fn(() => (Component) => Component),
// }));

// // Mock the react-router-dom module
// const mockNavigate = jest.fn();
// jest.mock('react-router-dom', () => ({
//   ...jest.requireActual('react-router-dom'),
//   useNavigate: () => mockNavigate,
// }));

// const mockBanners = [
//   // Provide mock data for your banners
// ];

// describe('Banner Page', () => {
//   test('Renders Banner page', () => {
//     render(
//       <Provider store={store}>
//         <IntlProvider locale="en" messages={{}}>
//           <Banner user={{ role: 'admin' }} banners={mockBanners} intl={{ formatMessage: jest.fn() }} />
//         </IntlProvider>
//       </Provider>
//     );

//     // Check if the component renders without errors
//     expect(screen.getByTestId('Banner')).toBeInTheDocument();

//     // Add more assertions based on your component's structure and behavior
//   });

//   test('handles button click and navigates to /addBanner', () => {
//     render(
//       <Provider store={store}>
//         <IntlProvider locale="en" messages={{}}>
//           <Banner user={{ role: 'admin' }} banners={mockBanners} intl={{ formatMessage: jest.fn() }} />
//         </IntlProvider>
//       </Provider>
//     );

//     // Find the button and simulate a click event
//     fireEvent.click(screen.getByText('app_add'));

//     // Check if the navigate function is called with the correct argument
//     expect(mockNavigate).toHaveBeenCalledWith('/addBanner');
//   });

//   // Add more tests as needed for different scenarios and edge cases
// });
