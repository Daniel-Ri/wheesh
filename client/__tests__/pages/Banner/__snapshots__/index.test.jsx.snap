// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Banner Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="Banner"
      >
        <div
          class="container"
        >
          <header>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="BackBtn"
              tabindex="0"
              type="button"
            >
              Kembali
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <h1>
              Manajemen Banner
            </h1>
          </header>
          <div
            class="button"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="AddBannerButton"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                data-testid="AddOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              <div
                class="message"
              >
                Tambah
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <table
            class="banners"
          >
            <tbody>
              <tr>
                <th>
                  Desktop
                </th>
                <th>
                  Mobile
                </th>
                <th />
              </tr>
            </tbody>
          </table>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="Banner"
    >
      <div
        class="container"
      >
        <header>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="BackBtn"
            tabindex="0"
            type="button"
          >
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <h1>
            Manajemen Banner
          </h1>
        </header>
        <div
          class="button"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="AddBannerButton"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
              data-testid="AddOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            <div
              class="message"
            >
              Tambah
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <table
          class="banners"
        >
          <tbody>
            <tr>
              <th>
                Desktop
              </th>
              <th>
                Mobile
              </th>
              <th />
            </tr>
          </tbody>
        </table>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
